### PR TITLE
Feature/default company

### PR DESCRIPTION
**Pull request by mujeeb-arshad**

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (mabdullahkhalil) [inline app/assets/javascripts/application.js]: no extra line

File: app/controllers/attendances_controller.rb [hunk 2]
```diff
@@ -15,12 +15,21 @@ def index
   def create
     @employees_todays_attendance = current_employee.todays_attendance_of_employee
     if @employees_todays_attendance.blank?
+      setting = Setting.all.first
```
Comment (mabdullahkhalil) [inline app/controllers/attendances_controller.rb]: shouldn't it be something like current_employee.company.settings or current_company.settings

File: app/controllers/attendances_controller.rb [hunk 3]
```diff
@@ -15,12 +15,21 @@ def index
...
+      today_start_time = setting.timings[Time.now.strftime('%A').downcase + '_start_time']
+      attendance_thresh = setting.attendance_time
+      if get_time_in_seconds(Time.now) > get_time_in_seconds(today_start_time.to_time) + attendance_thresh * 60
+        if current_employee.late_count.nil?
```
Comment (mabdullahkhalil) [inline app/controllers/attendances_controller.rb]: we can remove this check if late_count is set to zero when creating an employee.

File: app/controllers/events_controller.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
+class EventsController < ApplicationController
+  def index
+    @events = []
+    @event = Event.all
```
Comment (mabdullahkhalil) [inline app/controllers/events_controller.rb]: `current_company.events` or
`current_employee.company.events`

File: app/controllers/events_controller.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    @setting = Setting.all.first.holidays
```
Comment (mabdullahkhalil) [inline app/controllers/events_controller.rb]: `current_company.settings` or
`current_employee.company.settings`

File: app/controllers/events_controller.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    @events << @event
+    @events << @setting
+  end
+
+  def new
+    @event = Event.unscoped.new
```
Comment (mabdullahkhalil) [inline app/controllers/events_controller.rb]: `current_company.events or
current_employee.company.events`

File: app/controllers/events_controller.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  def home
+    @events = Event.all
+  end
+
+  def create
+    @event = Event.unscoped.new(event_params)
```
Comment (mabdullahkhalil) [inline app/controllers/events_controller.rb]: `

current_company.events or
current_employee.company.events
`

File: app/controllers/events_controller.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    @events << @event
+    @events << @setting
+  end
+
+  def new
+    @event = Event.unscoped.new
+  end
+
+  def home
+    @events = Event.all
```
Comment (mabdullahkhalil) [inline app/controllers/events_controller.rb]: `current_company.events or
current_employee.company.events`

File: app/controllers/events_controller.rb [hunk 9]
```diff
@@ -0,0 +1,61 @@
+class EventsController < ApplicationController
+  def index
+    @events = []
+    @event = Event.all
+    @setting = Setting.all.first.holidays
+    @events << @event
+    @events << @setting
```
Comment (mabdullahkhalil) [inline app/controllers/events_controller.rb]: you don't need to create an array. 
simply `@events` and `@settings` is enough

File: app/controllers/events_controller.rb [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+      flash[:success] = 'Event has been added!'
+      redirect_to index_events_path
+    else
+      flash[:danger] = 'Event has not been added!'
+      redirect_to menus_index_path
+    end
+  end
+
+  def edit
+    @event = Event.find(params[:id])
```
Comment (mabdullahkhalil) [inline app/controllers/events_controller.rb]: `current_company.events or current_employee.company.events`

File: app/controllers/events_controller.rb [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  def update
+    @event = Event.find(params[:id])
```
Comment (mabdullahkhalil) [inline app/controllers/events_controller.rb]: `current_company.events or current_employee.company.events`

File: app/controllers/events_controller.rb [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+      flash[:success] = 'Event has been updated!'
+      redirect_to index_events_path
+    else
+      flash[:danger] = 'Event has not been updated!'
+      redirect_to menus_index_path
+    end
+  end
+
+  def destroy
+    @event = Event.find(params[:id])
```
Comment (mabdullahkhalil) [inline app/controllers/events_controller.rb]: `current_company.events or current_employee.company.events`

File: app/controllers/menus_controller.rb [hunk 13]
```diff
@@ -1,10 +1,24 @@
...
-  def home
-    @companies = Company.all
+  def home   
   end
-end
\ No newline at end of file
+
+  def search_email
+    @email = params[:email]
+    if(@email != nil)
```
Comment (mabdullahkhalil) [inline app/controllers/menus_controller.rb]: `@email.nil?`

File: app/controllers/menus_controller.rb [hunk 14]
```diff
@@ -1,10 +1,24 @@
...
+      employees = Employee.where(email: params[:email])
```
Comment (mabdullahkhalil) [inline app/controllers/menus_controller.rb]: `current_company.employees or current_employee.company.employees`

File: app/controllers/menus_controller.rb [hunk 15]
```diff
@@ -1,10 +1,24 @@
...
+      @companies = []
+      employees.each do |employee|
+        company = Company.find_by(id: employee.company_id)
```
Comment (mabdullahkhalil) [inline app/controllers/menus_controller.rb]: `current_employee.company`

File: app/controllers/messages_controller.rb [hunk 16]
```diff
@@ -0,0 +1,11 @@
+class MessagesController < ApplicationController
+  
+  def index
+    @messages = current_company.messages
+    
+  end
+
+  def create
+    @message = Message.create!(params.require(:message).permit([:message, :employee_id, :company_id]))
```
Comment (mabdullahkhalil) [inline app/controllers/messages_controller.rb]: create a separate function for 
`params.require(:message).permit([:message, :employee_id, :company_id])`

File: app/controllers/sessions_controller.rb [hunk 17]
```diff
@@ -3,13 +3,24 @@ def new
     super
   end
   def create
+
```
Comment (mabdullahkhalil) [inline app/controllers/sessions_controller.rb]: extra line

File: app/controllers/sessions_controller.rb [hunk 18]
```diff
@@ -3,13 +3,24 @@ def new
...
     if(current_company.blank?)
       flash[:danger] = 'Invalid Company!'
       return redirect_to new_employee_session_path
     end
     employee = Employee.find_by(email: params[:employee][:email], company_id: current_company.id)
-    if employee.blank? || !employee.valid_password?(params[:employee][:password])
-      flash[:danger] = "Invalid password or email!"
+    if employee != nil
```
Comment (mabdullahkhalil) [inline app/controllers/sessions_controller.rb]: `employee.nil?`

File: app/controllers/sessions_controller.rb [hunk 19]
```diff
@@ -3,13 +3,24 @@ def new
...
+      if !employee.active
+        flash[:danger] = 'Can not log in, Employee is inactive!'
+      elsif employee.confirmation_token != nil
```
Comment (mabdullahkhalil) [inline app/controllers/sessions_controller.rb]: `employee.confirmation_token.nil?`

File: app/controllers/settings_controller.rb [hunk 20]
```diff
@@ -0,0 +1,69 @@
+class SettingsController < ApplicationController
+  def index
+    @settings = Setting.find_by(company_id: current_company.id)
```
Comment (mabdullahkhalil) [inline app/controllers/settings_controller.rb]: `current_employee.company.settings or current_company.settings`

File: app/controllers/settings_controller.rb [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+      if(@settings.working_days.class == String)
+        @settings.working_days = JSON.parse(@settings.working_days.gsub("'",'"').gsub('=>',':'))
+        @settings.timings = JSON.parse(@settings.timings.gsub("'",'"').gsub('=>',':'))
+      end
+      @settings
+    end
+  end
+
+  def edit
+    @settings = Setting.find_by(company_id: current_company.id)
```
Comment (mabdullahkhalil) [inline app/controllers/settings_controller.rb]: `current_employee.company.settings or current_company.settings`

File: app/controllers/settings_controller.rb [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+    if @settings != nil
```
Comment (mabdullahkhalil) [inline app/controllers/settings_controller.rb]: `@settings.nil?`

File: app/controllers/settings_controller.rb [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+      end
+      @settings
+    end
+  end
+
+  def update
+    days = [:monday, :tuesday, :wednesday, :thursday, :friday, :saturday, :sunday]
+    on_days = {}
+    days.each do |day|
+      if params[day].blank?
```
Comment (mabdullahkhalil) [inline app/controllers/settings_controller.rb]: `on_days[day] = false if params[day].blank?
on_days[day] = true if !params[day].blank?`

isn't this more convinient

File: app/controllers/settings_controller.rb [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+                'sunday_start_time', 'sunday_end_time']
+    timings = {}
+    daily_timings.each do |timing|
+      if params[timing].blank?
+        timings[timing] = '00:00 AM'
+      else
+        timings[timing] = params[timing]
+      end
+    end
+    @setting = Setting.find_by(company_id: current_company.id)
```
Comment (mabdullahkhalil) [inline app/controllers/settings_controller.rb]: `current_company.settings`

File: app/models/company.rb [hunk 25]
```diff
@@ -1,12 +1,23 @@
...
   has_many :tasks
   has_many :teams
   has_many :attendances
+  has_many :messages
+  has_one :setting
+  has_many :events
+  accepts_nested_attributes_for :setting
+
+  def self.current_id=(id)
+    Thread.current[:company_id] = id
```
Comment (mabdullahkhalil) [inline app/models/company.rb]: would `current_company.thread` work?

File: app/views/devise/registrations/new.html.erb [hunk 26]
```diff
@@ -1,63 +1,69 @@
...
+	      <%= company_form.label :description, "Company Description" %>
+	      <%= company_form.text_field :description,  class: 'form-control' %>
+	    </div>
+
+	    <div class="form-group">
+	      <%= company_form.label :subdomain, "Company Subdomain" %>
+	      <%= company_form.text_field :subdomain, class: 'form-control', id: 'company-subdomain', readonly: true %>
+	    </div>
+	      <%= company_form.fields_for :setting do |setting_form| %>
+	      <%= setting_form.hidden_field :timings, :value => ('{ "monday_start_time" => "00:00 AM", "monday_end_time" => "00:00 AM", "tuesday_start_time" => "00:00 AM", "tuesday_end_time" => "00:00 AM", "wednesday_start_time" => "00:00 AM", "wednesday_end_time" => "00:00 AM", "thursday_start_time" => "00:00 AM", "thursday_end_time" => "00:00 AM", "friday_start_time" => "00:00 AM", "friday_end_time" => "00:00 AM", "saturday_start_time" => "00:00 AM", "saturday_end_time" => "00:00 AM", "sunday_start_time" => "00:00 AM", "sunday_end_time" => "00:00 AM" }') %>
```
Comment (mabdullahkhalil) [inline app/views/devise/registrations/new.html.erb]: write a helper function to get this value

File: app/views/devise/registrations/new.html.erb [hunk 27]
```diff
@@ -1,63 +1,69 @@
...
+	      <%= setting_form.hidden_field :working_days, :value => ('{ "monday" => true, "tuesday" => true, "wednesday" => true, "thursday" => true, "friday" => true, "saturday" => true, "sunday" => true}') %>
```
Comment (mabdullahkhalil) [inline app/views/devise/registrations/new.html.erb]: write a helper function to get this value

File: app/views/events/index.html.erb [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+      datetime: new Date(<%= event.event_date.year%>, <%= event.event_date.month %> -1, <%= event.event_date.day%>, <%= event.event_date.strftime("%H") %>)
+    })
+  <% end %>
+  <% @events.last.each do |key,val| %>
+    events_arr.push({
+      title: "Public Holiday",
+      description: "This is a public holiday",
+      datetime: new Date(<%= Date.parse(@events.last[key]["holiday_date_#{key}"]).year %>, <%= Date.parse(@events.last[key]["holiday_date_#{key}"]).month %> -1, <%= Date.parse(@events.last[key]["holiday_date_#{key}"]).day%>, <%= Date.parse(@events.last[key]["holiday_date_#{key}"]).strftime("%H") %>)
+    })
+    <% if @events.last[key]["every_year_#{key}"] %>
```
Comment (mabdullahkhalil) [inline app/views/events/index.html.erb]: write a helper function for this

File: app/views/events/index.html.erb [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+  <% end %>
+  $('#calendar').eCalendar({
+  weekDays: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
+  months: ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
+  textArrows: {previous: '<', next: '>'},
+  eventTitle: 'Events',
+  url: '',
+  firstDayOfWeek: 0,
+  events : events_arr
+
```
Comment (mabdullahkhalil) [inline app/views/events/index.html.erb]: extra line

File: app/views/menus/home.html.erb [hunk 30]
```diff
@@ -1,21 +1,16 @@
...
+	  </div>
+	  <div class='form-group'>
+	    <%= submit_tag("Search", class:'btn btn-dark btn-block') %>
+	  </div>
+	<% end %>
+
+	<div id="_company-list">
+	  
+	</div>
+</div>
```
Comment (mabdullahkhalil) [inline app/views/menus/home.html.erb]: check for extra lines in the file

File: app/views/messages/index.html.erb [hunk 31]
```diff
@@ -0,0 +1,35 @@
+<div class="container">
+  <div class="title">Messenger</div>
+  <div class="window">
+    <div class="text" id="chat-output"></div>
+  </div>
+</div>
+<div class="container">
+  <div class="chatbox">
+    <%= form_for Message.new, remote: true do |msg| %>
```
Comment (mabdullahkhalil) [inline app/views/messages/index.html.erb]: make this in a variable in index action
`@message_new = current_company.message.new`

File: config/routes.rb [hunk 32]
```diff
@@ -1,11 +1,25 @@
...
   resources :teams
   resources :departments
   resources :attendances
 
+  get '/settings/new' => 'settings#new'
+  post '/settings/create' => 'settings#create', as: :create_settings
+  get '/settings/index' => 'settings#index', as: :index_settings
+  get '/settings/edit' => 'settings#edit', as: :edit_settings
+  patch '/settings/update' => 'settings#update', as: :update_settings
+  delete '/settings/destroy' => 'settings#destroy', as: :destroy_settings
```
Comment (mabdullahkhalil) [inline config/routes.rb]: `resources :settings`

File: config/routes.rb [hunk 33]
```diff
@@ -1,11 +1,25 @@
...
+
+  get '/events/index' => 'events#index', as: :index_events
+  get '/events/new' => 'events#new', as: :new_events
+  get '/events/home' => 'events#home', as: :home_events
+  post '/events/create' => 'events#create', as: :create_events
+  get '/events/:id/edit' => 'events#edit', as: :edit_events
+  patch '/events/:id/update' => 'events#update', as: :update_events
+  delete '/events/:id/destroy' => 'events#destroy', as: :destroy_events
```
Comment (mabdullahkhalil) [inline config/routes.rb]: `resources :events`

File: spec/support/controller_helpers.rb [hunk 34]
```diff
@@ -0,0 +1,11 @@
...
+  def sign_in(employee = double('employee'))
+    if employee.nil?
+      allow(request.env['warden']).to receive(:authenticate!).and_throw(:warden, { :scope => :employee })
+      allow(controller).to receive(:current_employee).and_return(nil)
+    else
+      allow(request.env['warden']).to receive(:authenticate!).and_return(employee)
+      allow(controller).to receive(:current_employee).and_return(employee)
+    end
+  end
+end
```
Comment (mabdullahkhalil) [inline spec/support/controller_helpers.rb]: check this

File: app/controllers/attendances_controller.rb (hunk 2 above)
Comment (mujeeb-arshad) [reply app/controllers/attendances_controller.rb]: already solved it in my current branch

File: app/controllers/attendances_controller.rb (hunk 3 above)
Comment (mujeeb-arshad) [reply app/controllers/attendances_controller.rb]: tried that gives errors that's why put this check

File: app/controllers/events_controller.rb (hunk 12 above)
Comment (mujeeb-arshad) [reply app/controllers/events_controller.rb]: scope defined already

File: app/controllers/menus_controller.rb (hunk 15 above)
Comment (mujeeb-arshad) [reply app/controllers/menus_controller.rb]: nope all companies required plus current_employee not set yes

File: app/controllers/messages_controller.rb (hunk 16 above)
Comment (mujeeb-arshad) [reply app/controllers/messages_controller.rb]: just calling once still be required?

File: app/controllers/settings_controller.rb (hunk 23 above)
Comment (mujeeb-arshad) [reply app/controllers/settings_controller.rb]: isn't both are fine?

File: app/models/company.rb (hunk 25 above)
Comment (mujeeb-arshad) [reply app/models/company.rb]: nope

File: app/views/devise/registrations/new.html.erb (hunk 26 above)
Comment (mujeeb-arshad) [reply app/views/devise/registrations/new.html.erb]: that's just default settings, won't need that I guess as per zain

File: app/views/events/index.html.erb (hunk 28 above)
Comment (mujeeb-arshad) [reply app/views/events/index.html.erb]: that's jquery has to be done here I guess

**Comment by mujeeb-arshad** [comment]
conflicts resolve

File: Gemfile [hunk 35]
```diff
@@ -24,6 +24,8 @@ gem 'sdoc', '~> 0.4.0', group: :doc
 gem 'select2-rails'
 gem 'sequenceid'
 gem 'uglifier', '>= 1.3.0'
+gem 'faye'
```
Comment (zainsaleem4338) [inline Gemfile]: specify version

File: Gemfile [hunk 36]
```diff
@@ -24,6 +24,8 @@ gem 'sdoc', '~> 0.4.0', group: :doc
...
+gem 'thin'
```
Comment (zainsaleem4338) [inline Gemfile]: specify version

File: app/controllers/events_controller.rb (hunk 7 above)
Comment (zainsaleem4338) [reply app/controllers/events_controller.rb]: why was this comment not resolved? reason?

File: app/controllers/events_controller.rb [hunk 37]
```diff
@@ -0,0 +1,61 @@
+class EventsController < ApplicationController
+  def index
```
Comment (zainsaleem4338) [inline app/controllers/events_controller.rb]: cancancan????
load_and_authorize

File: app/controllers/messages_controller.rb [hunk 38]
```diff
@@ -0,0 +1,10 @@
+class MessagesController < ApplicationController
+  def index
```
Comment (zainsaleem4338) [inline app/controllers/messages_controller.rb]: load_and_authorize

File: app/controllers/settings_controller.rb [hunk 39]
```diff
@@ -0,0 +1,69 @@
+class SettingsController < ApplicationController
+  def index
```
Comment (zainsaleem4338) [inline app/controllers/settings_controller.rb]: load_adn_authorize

File: app/models/employee.rb [hunk 40]
```diff
@@ -1,6 +1,7 @@
 require 'date'
 
 class Employee < ActiveRecord::Base
+  not_multitenant!
```
Comment (zainsaleem4338) [inline app/models/employee.rb]: why not multitenant ?

File: app/models/task.rb [hunk 41]
```diff
@@ -7,7 +7,6 @@ class Task < ActiveRecord::Base
   belongs_to :company
   belongs_to :project
   belongs_to :assignable, polymorphic: true
-  validates :complexity
```
Comment (zainsaleem4338) [inline app/models/task.rb]: verify with master or ask tehreem if we need this now?

**Pull request merged**